File: DataReconcile_Matlab/DRnonparamEqRobust.m
function reconciled_with_Kernels = DRnonparamEqRobust(Func, msrd_data, model_params, prior_vars, bandwidths)
% Func - указатель на функцию, описывающую зависимости между согласуемыми
% величинами
% msrd_data - вектор согласуемых результатов измерений вида (x_msdr1, x_msdr2, ..., x_msdrN ) 
% model_params - вектор параметров модели взаимосвязей, погрешности которых
% не заданы
% prior_vars - априорно заданные параметры неопределенности согласуемых
% величин
% bandwidths - вектор значений ширины 
    l = zeros([length(Func(msrd_data, model_params)), 1]);
    mu_and_l_start = mean(msrd_data, 2);
    for i = 1 : length(l)
        mu_and_l_start(length(mu_and_l_start)+1) = l(i);
    end

    % --- Robustness
    MAD = zeros([length(mean(msrd_data, 2)),1]);
    for jj = 1 : length(mean(msrd_data, 2))
        MAD(jj) = mad(msrd_data(jj,:), 1);
    end
    error_params = MAD .* 1.483;
    spn1 = eps; % Machine zero in Matlab
    % If calculated std is zero, then we use the external std value   
    for jj = 1 : length(mean(msrd_data, 2))
        if abs(error_params(jj)) < spn1
            error_params(jj) = prior_vars(jj);
        end
    end
    % ---


    mu_and_l_est = fsolve(@KernelsSystemToSolve, mu_and_l_start, [], ...
              Func, msrd_data, model_params, bandwidths, error_params);
    reconciled_with_Kernels =  mu_and_l_est(1:(length(mu_and_l_est)-length(l)));
end

function K = Kernel(x)
    K = exp(-x.^2 ./ 2)./sqrt(2*pi);
end

function mustbezeros = KernelsSystemToSolve(mu_and_l, DependFunc, ...
                                            msrd_data, model_params, ...
                                            bandwidths, msrd_vars)
    xNum = size(msrd_data, 1);
    n = size(msrd_data, 2);
    mustbezeros = zeros([xNum, 1]);
    if length(bandwidths) == 1
        bandwidths = zeros([1, xNum]) + bandwidths;
    end
    for j = 1 : xNum
        mean_msrd = mean(msrd_data(j,:));
        s_msrd = msrd_vars(j);
        res = zeros([1,n]);
        for i = 1:n
            denumerator = Kernel(( (msrd_data(j,:)-mean_msrd)./s_msrd - (msrd_data(j,i)-mu_and_l(j))./s_msrd )./bandwidths(j));
            denumerator = reshape(denumerator, 1, []); % делаем так, чтобы точно была строка!
            numerator = sum(msrd_data(j,:).*denumerator);
            if sum(denumerator) == 0
                res(i) = 0;
            else
                res(i) = numerator/sum(denumerator);
            end
        end
        mean_via_kerns = mean(res);
        imagx = zeros([xNum,1]); 
        imagx(j) = 1i*( mu_and_l(j)*10^(-100) + 10^(-101) );

        df_dx = imag(DependFunc(mu_and_l(1:xNum)+imagx, model_params))/imag(imagx(j));
        df_dx = reshape(df_dx,1, []);
        l_ = mu_and_l((xNum+1):length(mu_and_l));
        l_ = reshape(l_, 1, []);

        mustbezeros(j) = mean_msrd - mean_msrd - mean_via_kerns + mu_and_l(j) ...
                      + bandwidths(j)^2*(1/n)*s_msrd^2*sum(df_dx.*l_);
    end
    model_res = DependFunc(mu_and_l(1:xNum), model_params);
    for i = 1 : length(model_res)
        mustbezeros(length(mustbezeros)+1) = model_res(i);
    end
end

File: DataReconcile_Python/DRnonparamEqRobust.py
import numpy as np
from numpy import imag
from scipy.optimize import fsolve
from scipy.stats import median_abs_deviation

def kernel(x):
    """
    Computes the kernel value for the input x using a Gaussian kernel.
    Parameters:
        x (float or ndarray): Input value(s) for the kernel function.

    Returns:
        float or ndarray: Kernel value(s) computed as exp(-x^2 / 2) / sqrt(2 * pi).
    """
    return np.exp(-x**2 / 2) / np.sqrt(2 * np.pi)

# -----------------------------------------------------------------------------
# -----------------------------------------------------------------------------

def kernels_system_to_solve(mu_and_l, depend_func, msrd_data, model_params, bandwidths, msrd_vars):
    """
    Solves a system of equations using kernel-based estimation.

    Parameters:
        mu_and_l (ndarray): Array of unknown parameters (mu and lambda).
        depend_func (function): Dependency model function.
        msrd_data (ndarray): Measured data (n x xNum array).
        model_params (ndarray or list): model parameters.
        bandwidths (ndarray or float): Bandwidth parameter(s).
        msrd_vars (ndarray): Variance of measured data for each dimension.

    Returns:
        ndarray: A vector of values that need to be zero (mustbezeros).
    """
    xNum = msrd_data.shape[0]
    n = msrd_data.shape[1]
    mustbezeros = np.zeros(xNum, dtype=np.complex128)

    # Handle scalar bandwidth case
    if bandwidths.size == 1:
        bandwidths_ = np.full(xNum, bandwidths)
    else:
        bandwidths_ = bandwidths

    for j in range(xNum):
        mean_msrd = np.mean(msrd_data[j, :])

        # Compute median and scaled MAD
        mean_msrd = np.median(msrd_data[j, :])
        mad_value = median_abs_deviation(msrd_data[j, :], scale=1)
        s_msrd = mad_value * 1.483

        res = np.zeros(n)

        for i in range(n):
            # Compute kernel weights
            denumerator = kernel(((msrd_data[j, :] - mean_msrd) / s_msrd 
                                  - (msrd_data[j, i] - mu_and_l[j]) / s_msrd) / bandwidths_[j])
                
            numerator = np.sum(msrd_data[j, :] * denumerator)
            if np.sum(denumerator) == 0:
                res[i] = 0
            else:
                res[i] = numerator / np.sum(denumerator)

        mean_via_kerns = np.mean(res)

        # Numerical derivative using small imaginary perturbation
        imagx = np.zeros(xNum, dtype=np.complex128)
        imagx[j] = 1j * (mu_and_l[j] * 10 ** (-100) + 10 ** (-101))
        df_dx = imag(depend_func(mu_and_l[:xNum] + imagx, model_params)) / imag(imagx[j])

        # Update mustbezeros
        mustbezeros[j] = (mean_msrd - mean_msrd - mean_via_kerns + mu_and_l[j]
                          + bandwidths_[j]**2 * (1 / n) * s_msrd**2 * np.sum(df_dx * mu_and_l[xNum:]))

    # Evaluate the model residuals
    model_res = depend_func(mu_and_l[:xNum], model_params)
    mustbezeros = np.append(mustbezeros, model_res)

    return np.real(mustbezeros)

# -----------------------------------------------------------------------------
# -----------------------------------------------------------------------------

def DRnonparamEqRobust(func, msrd_data, model_params, prior_vars, bandwidths):
    """
    Reconciles measured data using kernel-based estimation.

    Parameters:
        func (function): Function describing dependencies between reconciled quantities.
        msrd_data (ndarray): Vector of measured results (n x xNum array).
        model_params (ndarray): Known model parameters.
        prior_vars (ndarray): A priori uncertainty parameters for reconciled quantities.
        bandwidths (ndarray): Vector of bandwidth values.

    Returns:
        ndarray: Reconciled measured data.
    """
    data_init_points = np.mean(msrd_data, axis=1)

    # Step 1: Compute l (a zero vector of the same length as func's output)
    l = np.zeros(len(func(data_init_points, model_params)))

    # Step 2: Initialize mu_and_l_start with the mean of msrd_data
    mu_and_l_start = np.hstack((data_init_points, l))  # Append zeros for l

    # Step 3: Define the system of equations and solve using fsolve
    def kernels_system_wrapper(mu_and_l):
        """
        Wrapper function for KernelsSystemToSolve to include all required parameters.
        """
        return kernels_system_to_solve(mu_and_l, func, msrd_data, model_params, bandwidths, prior_vars)

    # Solve the system of equations using fsolve
    mu_and_l_est = fsolve(kernels_system_wrapper, mu_and_l_start)

    # Step 4: Extract the reconciled measured data (excluding 'l' values)
    reconciled_with_kernels = mu_and_l_est[:len(mu_and_l_start) - len(l)]

    return reconciled_with_kernels


